### PR TITLE
Move funs in arguments to separate variables to avoid illegal patterns.

### PR DESCRIPTION
**Pull request by aggelgian**
e.g.
```
aggelgian@...:~/cuter$ ./cuter -r -v -p 2 -s 8 f2 mainfun '[fun(F) when is_function(F, 1) -> 42 end]'
Compiling f2.erl ... ok
Testing f2:mainfun/1 ...
f2:mainfun/1
    Spec:
        fun((fun((fun((integer()) -> integer())) -> integer())) -> ok)
    Types:


WARNING: Encountered an unsupported type while parsing the types in Core Erlang forms!
  Form: {type,608,map,any}

f2:mainfun(#Fun<erl_eval.6.128620087>) ... ok
xxxxxxxxxx
f2:mainfun(fun(_) -> 321 end) ... ok
xxxxxxxxx
A = fun(_) -> 7 end, f2:mainfun(fun(A) -> 654; (_) -> 321 end) ... ok
x

Solver Statistics ...
  - Solved models   : 2
  - Unsolved models : 20

No Runtime Errors Occured
```

### CODE REVIEW

File: src/cuter_pp.erl [hunk 1]
```diff
@@ -729,54 +729,91 @@ pp_input_minimal(Input, {M, F, _}) ->
...
+preprocess_fn(A, InFn) ->
+  case cuter_lib:is_lambda(A) of
+    false when is_list(A) ->
+      lists:flatten([preprocess_fn(E, InFn) || E <- A]);
+    false when is_tuple(A) ->
+      lists:flatten([preprocess_fn(E, InFn) || E <- tuple_to_list(A)]);
+    false ->
+      [];
+    true when InFn ->
+      [A] ++ preprocess_fn(A, false);
```
Comment (kostis) [inline src/cuter_pp.erl]: Perhaps make this `[A | preprocess_fn(A, false)];`  ?

**Comment by kostis** [comment]
Unfortunately, the translation is wrong -- I can explain in private.
The version I suggested with the `when` clauses needs to be used instead.

So the above example should to read something like:
```erl
F0 = fun(_) -> 7 end, f2:mainfun(fun(A) when A =:= F0 -> 654; (_) -> 321 end)
```
instead.

**Comment by aggelgian** [comment]
Updated to

```
aggelgian@bulldozer:~/cuter$ ./cuter -r -v -p 2 -s 8 f2 mainfun '[fun(F) when is_function(F, 1) -> 42 end]'
Compiling f2.erl ... ok
Testing f2:mainfun/1 ...
f2:mainfun/1
    Spec:
        fun((fun((fun((integer()) -> integer())) -> integer())) -> ok)
    Types:


WARNING: Encountered an unsupported type while parsing the types in Core Erlang forms!
  Form: {type,608,map,any}

f2:mainfun(#Fun<erl_eval.44.33894411>) ... ok
xxxxxxxxxxx
f2:mainfun(fun(_) -> 321 end) ... ok
xxxxxxxxx
A0 = fun(_) -> 7 end,f2:mainfun(fun(A) when A =:= A0 -> 654; (_) -> 321 end) ... ok
x

Solver Statistics ...
  - Solved models   : 2
  - Unsolved models : 21

No Runtime Errors Occured
```

**Comment by aggelgian** [comment]
We can now solve the `ff_f_g` testcase.
```bash
aggelgian@bulldozer:~/cuter$ ./cuter -v funs ff_f_g '[fun (F) when is_function(F, 1) -> 42 end, fun (_) -> 1 end, fun (_) -> 2 end]' -d 42 -p 2 -s 128 -pa /home/aggelgian/cuter/test/ftest/ebin
Testing funs:ff_f_g/3 ...
funs:ff_f_g/3
    Spec:
        fun((fun((funs@fun_i2i) -> integer()), funs@fun_i2i, funs@fun_i2i) -> ok)
    Types:
        funs@fun_i2i =>
            fun((integer()) -> integer())


WARNING: Encountered an unsupported type while parsing the types in Core Erlang forms!
  Form: {type,608,map,any}

funs:ff_f_g(#Fun<erl_eval.6.128620087>, #Fun<erl_eval.6.128620087>, #Fun<erl_eval.6.128620087>) ... ok
xxxxxxxxxx
funs:ff_f_g(fun(_) -> 123 end, fun(_) -> 4 end, fun(_) -> 4 end) ... ok
xxxxx
A0 = fun(10) -> 9; (_) -> 5 end, funs:ff_f_g(fun(A) when A =:= A0 -> 456; (_) -> 123 end, fun(_) -> 7 end, A0) ... RUNTIME ERROR
xxxx

Solver Statistics ...
  - Solved models   : 2
  - Unsolved models : 19

=== Inputs That Lead to Runtime Errors ===
#1 A0 = fun(10) -> 9; (_) -> 5 end, funs:ff_f_g(fun(A) when A =:= A0 -> 456; (_) -> 123 end, fun(_) -> 7 end, A0)
```

**Pull request merged**